keps/4400-workload-priority-preemption-policy/README.md

```diff
@@ -0,0 +1,137 @@
+
+Workloads using a `WorkloadPriorityClass` with `preemptionPolicy: Never` are subject to the additional quota constraint:
+
+- **Nominal Quota Limit**: They can only consume resources up to the ClusterQueue's nominal quota
+- **No Borrowing**: They cannot borrow resources from other ClusterQueues in the cohort  
+
+This prevents scenarios where non-preemptible workloads consume borrowed quota that cannot be reclaimed
+
+## Design Details
+
+### API Changes
+
+Extend the `WorkloadPriorityClass` type with a `preemptionPolicy` field:
+
+```golang
+// apis/kueue/v1beta1/workloadpriorityclass_types.go
+type WorkloadPriorityClass struct {
+    // ... existing fields ...
+    
+    // preemptionPolicy controls whether workloads using this priority class can be preempted.
+    // Valid values are "Always" (default) and "Never".
+    // "Always" means workloads can be preempted by higher priority workloads.
+    // "Never" means workloads cannot be preempted, but are restricted to nominal quota only.
+    // +kubebuilder:default=Always
+    // +kubebuilder:validation:Enum=Always;Never
+    // +optional
+    PreemptionPolicy *WorkloadPriorityClassPreemptionPolicy `json:"preemptionPolicy,omitempty"`
+}
+
+// WorkloadPriorityClassPreemptionPolicy describes a policy for if/when to preempt a workload.
+type WorkloadPriorityClassPreemptionPolicy string
+
+const (
+    // WorkloadPriorityClassPreemptionPolicyAlways means that workloads can always be preempted by
+    // other workloads with higher priority.
+    WorkloadPriorityClassPreemptionPolicyAlways WorkloadPriorityClassPreemptionPolicy = "Always"
+    
+    // WorkloadPriorityClassPreemptionPolicyNever means that workloads should never be preempted.
+    // Workloads with this policy are restricted to nominal quota and cannot
+    // borrow resources from other ClusterQueues.
+    WorkloadPriorityClassPreemptionPolicyNever WorkloadPriorityClassPreemptionPolicy = "Never"
+)
+```
+
+### Implementation Details
+
+1. **Add PreemptionPolicy field to WorkloadPriorityClass** - Add a field that can be `Always` or `Never` (extensible for future values)
+
+2. **Track NonPreemptibleUsage in clusterqueue_snapshot.go** - Calculate current usage by non-preemptible workloads for each FlavorResource
+
+3. **Update flavorassigner.go/fitsResourceQuota** - If a workload is non-preemptible, check that it doesn't exceed the non-preemptible quota constraint
+
+4. **Update preemption.go/findCandidates** - Skip non-preemptible workloads when selecting preemption candidates
+
+5. **Define IsNonPreemptible helper** - `getWorkloadPriorityClass(w).preemptionPolicy == Never`
+
+### Interaction with ClusterQueue Preemption Policies
+
+The preemption policy hierarchy works as follows:
+
+1. **ClusterQueue policy is checked first** - If the ClusterQueue disables preemption, no workloads in that queue can be preempted regardless of their WorkloadPriorityClass. However, the quota constraints for non-preemptible workloads still apply.
```

Comment:
This ClusterQueue configuration defines which workloads the ClusterQueue can preempt (including its own workloads, via [WithinClusterQueue](https://github.com/kubernetes-sigs/kueue/blob/22e4502335b0116300b3dcc63d237cf8b7ac759a/apis/kueue/v1beta1/clusterqueue_types.go#L470-L472)). We do not currently have a configuration to prevent other ClusterQueues from preempting workloads in the ClusterQueue defining the policy, except configuration the BorrowingLimit to 0 (since nominal quota can't ever be preempted)